File: leetcode/palindrome_linked_list.js
/* Given a singly linked list, determine if it is a palindrome.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    var slow = head, fast = head;
    for (var prev = null; fast && fast.next; ) {
        fast = fast.next.next;
        var temp = slow.next;
        slow.next = prev;
        prev = slow;
        slow = temp;
    }
    if (fast)
        slow = slow.next;
    while (prev && prev.val === slow.val) {
        prev = prev.next;
        slow = slow.next;
    }
    return prev === null;
};

File: lintcode/palindrome_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a boolean
    def isPalindrome(self, head):
        slow, fast, prev = head, head, None
        while fast and fast.next:
            fast, prev, prev.next, slow = fast.next.next, slow, prev, slow.next
        if fast: slow = slow.next
        while prev and prev.val == slow.val:
            prev, slow = prev.next, slow.next
        return not prev
